src/Raven.Client/Documents/Session/AsyncDocumentSession.cs

```diff
@@ -139,10 +139,16 @@ public async Task SaveChangesAsync(CancellationToken token = default(Cancellatio
 
                 if (NoTracking)
                     throw new InvalidOperationException($"Cannot execute '{nameof(SaveChangesAsync)}' when entity tracking is disabled in session.");
-
-                await RequestExecutor.ExecuteAsync(command, Context, SessionInfo, token).ConfigureAwait(false);
-                UpdateSessionAfterSaveChanges(command.Result);
-                saveChangesOperation.SetResult(command.Result);
+                try
+                {
+                    await RequestExecutor.ExecuteAsync(command, Context, SessionInfo, token).ConfigureAwait(false);
+                    UpdateSessionAfterSaveChanges(command.Result);
+                    saveChangesOperation.SetResult(command.Result);
+                }
+                finally
+                {
+                    command.Result?.Results?.Dispose();
```

Comment:
Can we move this to the `BatchCommand.Dispose` ? That is called anyway, no?

One thing that worries me here, though, is the `SetResult`. IIRC, we are holding the blittable in the session, and they are the same as the one we have here, no?
If we can `SaveChanges` twice, won't that break?

Reply:
Good questions, 
I'm not sure if BatchCommand.Dispose will dispose of the results it seems to dispose of the commands as blittable (in the base class) while here the command is serialized to a C# object so I'm not sure those are the same instances, need to check.
I also wasn't sure about SetResult I ran slow tests and they passed so either we don't have this in our tests or the results are been cloned, either way, we need to test this scenario explicitly.
Regarding save changes twice we could clean the session commands in the first SaveChanges.

Reply:
1. After testing this, save changes twice seems safe.
2. indeed the commands are results are not related as I expected, but we could probably move the dispose of the result into the dispose of the batch command.